core/bindings.rs

```diff
@@ -0,0 +1,88 @@
+use crate::isolate::Buf;
+use crate::libdeno::PinnedBuf;
+use futures::Future;
+use std::fmt;
+use std::io;
+
+pub type BindingResult<T> = std::result::Result<T, BindingError>;
+
+#[derive(Debug)]
+pub struct BindingError {
+  repr: Repr,
+}
+
+#[derive(Debug)]
+enum Repr {
+  Simple(String),
+  IoErr(io::Error),
+}
+
+pub fn new_binding_error(msg: String) -> BindingError {
+  BindingError {
+    repr: Repr::Simple(msg),
+  }
+}
+
+impl fmt::Display for BindingError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self.repr {
+      Repr::Simple(ref err_str) => f.pad(err_str),
+      Repr::IoErr(ref err) => err.fmt(f),
+    }
+  }
+}
+
+impl std::error::Error for BindingError {
+  fn description(&self) -> &str {
+    match self.repr {
+      Repr::Simple(ref msg) => msg.as_str(),
+      Repr::IoErr(ref err) => err.description(),
+    }
+  }
+
+  fn cause(&self) -> Option<&dyn std::error::Error> {
+    match self.repr {
+      Repr::Simple(ref _msg) => None,
+      Repr::IoErr(ref err) => Some(err),
+    }
+  }
+}
+
+impl From<io::Error> for BindingError {
+  #[inline]
+  fn from(err: io::Error) -> Self {
+    Self {
+      repr: Repr::IoErr(err),
+    }
+  }
+}
+
+pub type BindingOpAsyncFuture =
+  Box<dyn Future<Item = Buf, Error = BindingError> + Send>;
+
+pub enum BindingOpSyncOrAsync {
+  Sync(Buf),
+  Async(BindingOpAsyncFuture),
+}
+
+pub type BindingOpResult = BindingResult<BindingOpSyncOrAsync>;
+
+/// Dispatch funciton type
+/// base is a placeholder value for now not sure what we want to use there
+pub type BindingOpDispatchFn =
+  fn(is_sync: bool, data: &[u8], zeroCopy: Option<PinnedBuf>)
+    -> BindingOpResult;
```

Comment:
I would prefer if we can use the same dispatch function used in isolate.rs...

Reply:
Need the `is_sync` value so we know what kind of return is expected.

Reply:
Yes, perhaps the normal dispatch function should be changed to also take is_sync.